### ISSUE TITLE
**Explore:** Cannot delete App when reconciliation is stuck

### ISSUE THREAD

**Issue by alexbrand** [carvel-dev/kapp-controller#69]
**What steps did you take:**

1. Create an App with the following definition:
```
apiVersion: kappctrl.k14s.io/v1alpha1
kind: App
metadata:
  name: simple-app
spec:
  syncPeriod: 10s
  serviceAccountName: default
  fetch:
    - inline:
        paths:
          deploy.yaml: |
            apiVersion: apps/v1
            kind: Deployment
            metadata:
              labels:
                app: bad
              name: bad
            spec:
              selector:
                matchLabels:
                  app: bad
              template:
                metadata:
                  labels:
                    app: bad
                spec:
                  containers:
                  - image: bad-image
                    name: bad-image
  template:
    - ytt: {}
  deploy:
    - kapp: {}
```

2. App gets stuck reconciling because the image points to a non-existing container image
3. Attempt to delete the app using `kubectl delete app simple-app`

**What happened:**
The app is stuck reconciling

**What did you expect:**
The app is deleted

**Comment by cppforlife** [comment]
hey @alexbrand, by default kapp deploy times out after 15min (default value for --wait-timeout in kapp deploy cmd). after that kapp-controller would try to reconcile and see that App was marked for deletion so it will kick off kapp delete and ultimately delete App CR.

there is "break glass" option for interrupting a deploy:
- issue delete command
- go to edit resource and add `spec.canceled: true` (this forces kapp-controller to exit currently running deploy run)
- remove spec.canceled (to allow kapp-controller to enqueue pending reconciliation actions)
- resource should get deleted now

another thing that some folks have done is:

```yaml
deploy:
- kapp:
    delete:
      rawOptions: ["--wait-timeout=5m"]
```

but that just reduces timeout.

it would be interesting to improve behaviour of kapp-controller in this area

- may be it's possible for kapp to be a little smarter in detecting when Deployment resource is "not going to recover" (in theory in this case it could if image suddently appears in the registry, that's why it's not considered terminal).
- or may be in addition to watching App CR for spec.canceled, we should also react to marking for deletion and exit current reconciliation

**Comment by aaronshurley** [comment]
Marking this issue as `Explore` so that we can spend some time defining a solution (and desired UX) for this.

**Comment by ewrenn8** [comment]
@cppforlife I am leaning towards the second improvement of exiting the regular reconciliation when we notice a deletion event has occurred as a solution to this issue. Any more thoughts you have on this or is that an alright path forward?

**Comment by joe-kimmel-vmw** [comment]
fixed by #569
